pcsx2/Recording/InputRecording.h

```diff
@@ -19,38 +19,78 @@
 
 
 #ifndef DISABLE_RECORDING
-enum INPUT_RECORDING_MODE
-{
-	INPUT_RECORDING_MODE_NONE,
-	INPUT_RECORDING_MODE_RECORD,
-	INPUT_RECORDING_MODE_REPLAY,
-};
-
 class InputRecording
 {
 public:
-	InputRecording() {}
-	~InputRecording(){}
+	// Main handler for ingesting input data and either saving it to the recording file (recording)
+	// or mutating it to the contents of the recording file (replaying)
+	void ControllerInterrupt(u8 &data, u8 &port, u16 &BufCount, u8 buf[]);
 
-	void ControllerInterrupt(u8 &data, u8 &port, u16 & BufCount, u8 buf[]);
+	// The running frame counter for the input recording
+	u32 GetFrameCounter();
 
-	void RecordModeToggle();
+	InputRecordingFile &GetInputRecordingData();
+
+	// The internal PCSX2 g_FrameCount value on the first frame of the recording
+	u32 GetStartingFrame();
 
-	INPUT_RECORDING_MODE GetModeState();
-	InputRecordingFile & GetInputRecordingData();
+	void IncrementFrameCounter();
+
+	// DEPRECATED: Slated for removal 
+	// If the current frame contains controller / input data
 	bool IsInterruptFrame();
 
-	void Stop();
+	// If there is currently an input recording being played back or actively being recorded
+	bool IsRecordingActive();
+
+	// Whether or not the recording's initial save state has yet to be loaded or saved and 
+	// the rest of the recording can be initialized
+	// This is not applicable to recordings from a "power-on" state
+	bool IsSavestateInitializing();
+
+	// If there is currently an input recording being played back
+	bool IsRecordingReplaying();
+
+	// String representation of the current recording mode to be interpolated into the title
+	wxString RecordingModeTitleSegment();
+
+	// Switches between recording and replaying the active input recording file
+	void RecordModeToggle();
+
+	// Mark the recording's initial savestate as having been loaded or saved successfully
+	void SavestateInitialized();
+
+	// Set the running frame counter for the input recording to an arbitrary value
+	void SetFrameCounter(u32 newFrameCounter);
+
+	// Store the starting internal PCSX2 g_FrameCount value
+	void SetStartingFrame(u32 newStartingFrame);
+	
+	/// Functions called from GUI
+
+	// Create a new input recording file
 	void Create(wxString filename, bool fromSaveState, wxString authorName);
+	// Play an existing input recording from a file
 	void Play(wxString filename, bool fromSaveState);
+	// Stop the active input recording
+	void Stop();
 
 private:
-	InputRecordingFile InputRecordingData;
-	INPUT_RECORDING_MODE state = INPUT_RECORDING_MODE_NONE;
+	enum class InputRecordingMode
+	{
+		NoneActive,
```

Comment:
Just copying this comment here onto the PR level so it's a bit easier to find, fyi @sonicfind 
> If I may: as you know, I'm implementing multitap support for input recording. One of the essential features is the ability to precisely choose a specific pad to record from, so I plan on giving each pad its own inputRecordingMode value. It could be good to rename this as NotActive instead for futureproofing.

After some (brief) thought, I'm not sure if that is the right way to go, though I definitely need more understanding around the multitap implementation (admittedly havn't looked at it too much since things are blocked by in-progress branches).  

My concern is currently the `InputRecordingMode` is used as a way to globally indicate recording status:
1. Is there going ever be a situation where one pad is recording, while another is replaying?  
2. Or is it just a case of selecting which pads will be recording/replaying from and their status' will always match?  

If it's the second situation, then that sounds fine because it would be trivial to iterate through the pad's status, and find if: any are replaying, any are recording, all are inactive.

Reply:
Being able to have different pads set to differing recording modes at one time is an essential part of TASing multiplayer. Trying to TAS two or more controllers at the same while having no ability to save each controller's inputs separately is a grueling task or, in some instances, impossible. As a personal example: TASing VS. Play in Gitaroo Man. Believe me, I tried.

Each pad will have its own InputRecordingMode value, however, there will still be a single main value "state" held for Shift+R toggle purposes. This value would also still be used as a means of determining whether or not a TAS file is loaded.

I'm including the snippets of my implementation that pertains to this + some other slightly-relevant QoL changes. You can look over them and then make a decision on whether or not to go with it.
```
// References to PadData or virtualPad objects are additions stemming
// from pcsx2-rr branches that haven't been merged/pulled yet

//===============================================//
//                InputRecording.h               //
//===============================================//
class InputRecording
{
public:
	// It was necessary to add more direct methods that perform
	// the task setting all active pads to record or replay mode
	// Initially done for InputRecordingControls.cpp
	// Sets all active pads to Record Mode
	void SetToRecordMode();

	// Sets all active pads to Replay Mode
	void SetToReplayMode();

	// Switches between recording and replaying the active input recording file
	void RecordModeToggle();

	// Switches between recording and replaying inputs from the chosen pad
	void RecordModeToggle(const u8 port, const u8 slot);

	// Prints all pad record mode states to the log window
	void DisplayActiveRecordModeStates() const;
private:
	InputRecordingMode state = InputRecording::InputRecordingMode::NotActive;

	// Holds the data, virtualpad controller, input mode, and fseek displacement for each pad
	struct RecordingPad
	{
		// Future: Controller Data
		PadData* padData;
		// Future: VirtualPad
		VirtualPad* virtualPad;
		// Pad Recording Mode
		InputRecordingMode inputState;
		// Fseek Displacement
		u32 seekOffset;
		RecordingPad();
		// Future
		~RecordingPad();
	};
	RecordingPad pads[2][4];
}

//===============================================//
//               InputRecording.cpp              //
//===============================================//

InputRecording::RecordingPad::RecordingPad()
{
	// Future
	virtualPad = nullptr;
	// Future
	padData = new PadData();
	inputState = InputRecordingMode::NotActive;
}

// Future
InputRecording::RecordingPad::~RecordingPad()
{
	delete padData;
}

//
// Overall has the most visible alterations
//
void InputRecording::ControllerInterrupt(const u8& data, const u8& port, const u8& slot, const u16& bufCount, u8(&buf)[512])
{
	RecordingPad& selectedPad = pads[port][slot];
	// We are only interested in reading from the buffer when it contains the controller data.
	// The first byte in the buffer is used to indicate this.
	// A flag is set(fInterruptFrame) so that future interrupts will either be processed or skipped.
	if (bufCount == 1)
	{
        fInterruptFrame = data == READ_DATA_AND_VIBRATE_FIRST_BYTE;
		return;
	}
	else if (bufCount == 2)
	{
        	// This is an extra sanity check, if this byte is different, then we unset the flag
		// and skip processing the buffer for this particular interrupt routine.
		if (buf[bufCount] != READ_DATA_AND_VIBRATE_SECOND_BYTE)
		{
			fInterruptFrame = false;
		}
		return;
	}
	else if (bufCount >= 3)
	{
		u8& bufVal = buf[bufCount];
		// For convenience sake, since the controller data starts at bufCount=3, we treat that as the first index. 
		const u16 bufIndex = bufCount - 3;

		// If the controller is in replay mode, there should be NO modifications to the inputs
		// Grab the byte from the movie file and overwrite whatever the [Virtual]Pad is inputting
		if (selectedPad.inputState == InputRecordingMode::Replaying)
		{
			u8 tmp = 0;
			//
			// Added seekOffset parameter so that there is no need to recalculate the needed offset for every
			// pad read based off how many pads are listed before it.
			//
			if (inputRecordingData.ReadKeyBuffer(tmp, g_FrameCount, port, slot, selectedPad.seekOffset, bufIndex))
			{
				// Overwrite value originally provided by the PAD plugin
				bufVal = tmp;
			}
			// Update controller data state for future VirtualPad / logging usage.
			selectedPad.padData->UpdateControllerData(bufIndex, bufVal);

			// If we have reached the end of the pad data, log it out
			if (bufIndex == PadData::END_INDEX_CONTROLLER_BUFFER)
			{
				selectedPad.padData->LogPadData(port, slot);
				// As well as re-render the virtual pad UI, if applicable
				// - Don't render if it's minimized
				if (selectedPad.virtualPad && selectedPad.virtualPad->IsShown() && !selectedPad.virtualPad->IsIconized())
				{
					selectedPad.virtualPad->Redraw();
				}
			}
		}
		else
		{
			// Update controller data state for future VirtualPad / logging usage.
			selectedPad.padData->UpdateControllerData(bufIndex, bufVal);

			if (selectedPad.virtualPad && selectedPad.virtualPad->IsShown())
			{
				// If the VirtualPad updated the PadData, we have to update the buffer
				// before committing it to the recording / sending it to the game
				// - Do not do this if we are in replay mode!
				if (selectedPad.virtualPad->UpdateControllerData(bufIndex, selectedPad.padData))
				{
					bufVal = selectedPad.padData->PollControllerData(bufIndex);
				}
			}

			// If we have reached the end of the pad data, log it out
			if (bufIndex == PadData::END_INDEX_CONTROLLER_BUFFER) {
				selectedPad.padData->LogPadData(port, slot);
				// As well as re-render the virtual pad UI, if applicable
				// - Don't render if it's minimized
				if (selectedPad.virtualPad && selectedPad.virtualPad->IsShown() && !selectedPad.virtualPad->IsIconized())
				{
					selectedPad.virtualPad->Redraw();
				}
			}

			// Finally, commit the byte to the movie file if we are recording
			if (selectedPad.inputState == InputRecordingMode::Recording)
			{
				inputRecordingData.SetTotalFrames(g_FrameCount);
				//
				// Added seekOffset parameter so that there is no need to recalculate the needed offset for every
				// pad write based off how many pads are listed before it.
				//
				inputRecordingData.WriteKeyBuffer(g_FrameCount, port, slot, selectedPad.seekOffset, bufIndex, bufVal);
			}
		}
	}
}

bool InputRecording::IsRecordingReplaying() const
{
	for (u8 port = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			if (pads[port][slot].inputState == InputRecordingMode::Replaying)
			{
				return true;
			}
		}
	}
	return false;
}

void InputRecording::SetToRecordMode()
{
	state = InputRecordingMode::Recording;
	// Set active VirtualPads to record mode
	for (u8 port = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			if (pads[port][slot].inputState == InputRecordingMode::Replaying)
			{
				pads[port][slot].virtualPad->SetReadOnlyMode(false);
				pads[port][slot].inputState = InputRecordingMode::Recording;
			}
		}
	}
	if (inputRecordingData.GetPadCount() == 1)
	{
		recordingConLog("[REC]: Record mode ON.\n");
	}
	else
	{
		recordingConLog("[REC]: All pads set to Record mode.\n");
	}
}

void InputRecording::SetToReplayMode()
{
	state = InputRecordingMode::Replaying;
	// Set active VirtualPads to readOnly mode
	for (u8 port = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			if (pads[port][slot].inputState == InputRecordingMode::Recording)
			{
				pads[port][slot].virtualPad->SetReadOnlyMode(true);
				pads[port][slot].inputState = InputRecordingMode::Replaying;
			}
		}
	}
	if (inputRecordingData.GetPadCount() == 1)
	{
		recordingConLog("[REC]: Replay mode ON.\n");
	}
	else
	{
		recordingConLog("[REC]: All pads set to Replay mode.\n");
	}
}

void InputRecording::RecordModeToggle()
{
	if (state == InputRecordingMode::Replaying)
	{
		SetToRecordMode();
	}
	else if (state == InputRecordingMode::Recording)
	{
		SetToReplayMode();
	}
}

void InputRecording::RecordModeToggle(const u8 port, const u8 slot)
{
	if (inputRecordingData.GetPadCount() == 1)
	{
		RecordModeToggle();
	}
	else if (pads[port][slot].inputState == InputRecordingMode::NotActive)
	{
		recordingConLog(wxString::Format(L"[REC]: Pad %d%c is not in use.\n", port + 1, 'A' + slot));
	}
	else
	{
		if (pads[port][slot].inputState == InputRecordingMode::Replaying)
		{
			// Set chosen VirtualPad to record mode
			pads[port][slot].virtualPad->SetReadOnlyMode(false);
			pads[port][slot].inputState = InputRecordingMode::Recording;
			state = InputRecordingMode::Recording;
			recordingConLog(wxString::Format(L"[REC]: Pad %d%c set to Record Mode.\n", port + 1, 'A' + slot));
		}
		else
		{
			// Set chosen VirtualPad to readOnly mode
			pads[port][slot].virtualPad->SetReadOnlyMode(true);
			pads[port][slot].inputState = InputRecordingMode::Replaying;
			state = InputRecordingMode::Replaying;
			recordingConLog(wxString::Format(L"[REC]: Pad %d%c set to Replay Mode.\n", port + 1, 'A' + slot));
		}
		DisplayAllRecordModeStates()
	}
}

bool InputRecording::DisplayActiveRecordModeStates() const
{
	for (u8 port = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			if (pads[port][slot].inputState == InputRecordingMode::Recording)
			{
				recordingConLog(wxString::Format(L"[REC]: Pad %d%c: Recording\n", port + 1, 'A' + slot));
				record = true;
			}
			else if (pads[port][slot].inputState == InputRecordingMode::Replaying)
			{
				recordingConLog(wxString::Format(L"[REC]: Pad %d%c: Replaying\n", port + 1, 'A' + slot));
				replay = true;
			}
		}
	}
}

void InputRecording::Stop()
{
	state = InputRecordingMode::NotActive;
	if (inputRecordingData.Close())
	{
		recordingConLog(L"[REC]: InputRecording Recording Stopped.\n");
		for (u8 port = 0; port < 2; port++)
		{
			for (u8 slot = 0; slot < 4; slot++)
			{
				if (pads[port][slot].inputState != InputRecordingMode::NotActive)
				{
					pads[port][slot].virtualPad->SetReadOnlyMode(false);
					pads[port][slot].inputState = InputRecordingMode::NotActive;
					pads[port][slot].seekOffset = 0;
				}
			}
		}
	}
}

/* Used in InputRecording::Create

	state = InputRecordingMode::Recording;
	// Ensure active VirtualPads aren't set to readOnly mode
	for (u8 port = 0, padsUsed = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			// New InputRecordingData method accessing a 2D boolean array of active pad values
			// True if active
			if (inputRecordingData.IsSlotUsed(port, slot))
			{
				pads[port][slot].virtualPad->SetReadOnlyMode(false);
				pads[port][slot].inputState = InputRecordingMode::Recording;
				pads[port][slot].seekOffset = 18 * padsUsed;
				padsUsed++;
			}
		}
	}
	
*/

/* Used in InputRecording::Play

	state = InputRecordingMode::Replaying;
	// Set active VirtualPads to readOnly mode
	for (u8 port = 0, padsUsed = 0; port < 2; port++)
	{
		for (u8 slot = 0; slot < 4; slot++)
		{
			// New InputRecordingData method accessing a 2D boolean array of active pad values
			// True if active
			if (inputRecordingData.IsSlotUsed(port, slot))
			{
				pads[port][slot].virtualPad->SetReadOnlyMode(true);
				pads[port][slot].inputState = InputRecordingMode::Replaying;
				pads[port][slot].seekOffset = 18 * padsUsed;
				padsUsed++;
			}
		}
	}
*/

//===============================================//
//           InputRecordingControls.cpp          //
//===============================================//

// It could be a good idea to swap the conditional statements around so it evaluates the frame counts first
// for possibly quicker short circuiting

void InputRecordingControls::FrameAdvance()
{
	if (g_InputRecording.IsRecordingReplaying() && g_InputRecording.GetFrameCounter() >= g_InputRecording.GetInputRecordingData().GetTotalFrames())
	{
		g_InputRecording.SetToRecordMode();
		return;
	}
	frameAdvanceMarker = g_InputRecording.GetFrameCounter();
	frameAdvancing = true;
	Resume();
}

void InputRecordingControls::Resume()
{
	if (g_InputRecording.IsRecordingReplaying() && g_InputRecording.GetFrameCounter() >= g_InputRecording.GetInputRecordingData().GetTotalFrames())
	{
		g_InputRecording.SetToRecordMode();
		return;
	}
	pauseEmulation = false;
	resumeEmulation = true;
}

void InputRecordingControls::TogglePause()
{
	if (pauseEmulation && g_InputRecording.IsRecordingReplaying() && g_InputRecording.GetFrameCounter() >= g_InputRecording.GetInputRecordingData().GetTotalFrames())
	{
		g_InputRecording.SetToRecordMode();
		return;
	}
	pauseEmulation = !pauseEmulation;
	resumeEmulation = !pauseEmulation;
}
```

Reply:
That justification makes sense to me, thanks for that / and the in-progress code.  I'll make the wording adjustment as no matter the outcome, `NotActive` is better naming.